api/tools/src/main/java/org/pmiops/workbench/tools/elastic/BigQueryOmopIndexer.java

```diff
@@ -0,0 +1,325 @@
+      .hasArg()
+      .build();
+  private static Option deleteIndicesOpt = Option.builder()
+      .longOpt("delete-indices")
+      .desc("If specified, deletes existing indices of the same name before re-importing")
+      .build();
+  private static Options createOptions = new Options()
+      .addOption(queryProjectIdOpt)
+      .addOption(esBaseUrlOpt)
+      .addOption(cdrVersionOpt)
+      .addOption(cdrBigQueryDatasetOpt)
+      .addOption(scratchBigQueryDatasetOpt)
+      .addOption(scratchGcsOpt)
+      .addOption(inverseProbOpt)
+      .addOption(deleteIndicesOpt);
+
+  private static final Logger log = Logger.getLogger(BigQueryOmopIndexer.class.getName());
+
+  // Recommended document type for all indices is "_doc" (types are being deprecated):
+  // https://www.elastic.co/guide/en/elasticsearch/reference/current/removal-of-types.html
+  private static final String ES_INDEX_TYPE = "_doc";
+  private static final RequestOptions ES_OPTS = RequestOptions.DEFAULT;
+
+  private static final int BATCH_SIZE = 500;
+
+  private RestHighLevelClient client;
+
+  private void createIndex(CommandLine opts) throws IOException, InterruptedException {
+    Preconditions.checkArgument(
+        !(opts.hasOption(scratchGcsOpt.getLongOpt()) ^
+            opts.hasOption(scratchBigQueryDatasetOpt.getLongOpt())),
+        "Must provide both or provide neither of scratch BigQuery and scratch GCS locations");
+    int inverseProb = 1;
+    if (opts.hasOption(inverseProbOpt.getLongOpt())) {
+      inverseProb = Integer.parseInt(opts.getOptionValue(inverseProbOpt.getLongOpt()));
+    }
+
+    // Initialize the index.
+    String personIndex = opts.getOptionValue("cdr-version") + "_person";
+    if (opts.hasOption(deleteIndicesOpt.getLongOpt()) &&
+        client.indices().exists(new GetIndexRequest().indices(personIndex), ES_OPTS)) {
+      client.indices().delete(new DeleteIndexRequest(personIndex), ES_OPTS);
+      log.info("deleted existing index: " + personIndex);
+    }
+    try {
+      client.indices().create(
+          new CreateIndexRequest(personIndex)
+              .mapping(ES_INDEX_TYPE, ImmutableMap.of("properties", ElasticDocument.PERSON_SCHEMA)),
+          ES_OPTS);
+    } catch (ElasticsearchStatusException e) {
+      if (e.status().getStatus() != 400) {
+        throw e;
+      }
+      log.warning(
+          "400 status on index creation, assuming conflict and ignoring: " + e.getMessage());
+    }
+
+    // Generate JSON documents from the CDR for Elasticsearch.
+    // Note: Here, BigQuery uses default credentials for the given environment.
+    BigQuery bq = BigQueryOptions.newBuilder()
+        .setProjectId(opts.getOptionValue(queryProjectIdOpt.getLongOpt()))
+        .build()
+        .getService();
+    String cdrDataset = opts.getOptionValue(cdrBigQueryDatasetOpt.getLongOpt());
+    int totalSampleSize = getPersonCount(bq, cdrDataset, inverseProb);
+    String personSQL = getPersonBigQuerySQL(cdrDataset, inverseProb);
```

Comment:
can the count not be gotten from the result of personSql?

Reply:
I'd assumed I couldn't get the total row count when I did an extraction to a table; I was wrong. Done - thanks.